### PR TITLE
feat(optimizer): support ApplyJoinRule one side push down optimization

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Currently, `ApplyJoinRule` always pushes down apply to the both sides of the join.
- This PR support one side push down optimization for `ApplyJoinRule` which means if any one side is not dependent on the apply and then we can push the join to the other side.
- This optimization can eliminate the unnecessary join generated by `ApplyJoinRule` before.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
[Unnesting Arbitrary Queries](https://btw-2015.informatik.uni-hamburg.de/res/proceedings/Hauptband/Wiss/Neumann-Unnesting_Arbitrary_Querie.pdf)

### CODE REVIEW

File: src/frontend/src/optimizer/rule/apply_join.rs [hunk 1]
```diff
@@ -42,17 +52,320 @@ impl Rule for ApplyJoinRule {
...
+                LogicalFilter::create(
+                    new_join.into(),
+                    Condition {
+                        conjunctions: other_condition,
+                    },
+                )
+            }
+            JoinType::RightSemi | JoinType::RightAnti | JoinType::Unspecified => unreachable!(),
+        }
+    }
```
Comment (likg227) [inline src/frontend/src/optimizer/rule/apply_join.rs:242]:
```
                LogicalFilter::create(
                    new_join.into(),
                    Condition {
                        conjunctions: other_condition,
                    },
                )
```
is enough?

File: src/frontend/src/optimizer/rule/apply_join.rs [hunk 2]
```diff
@@ -42,17 +52,320 @@ impl Rule for ApplyJoinRule {
...
+                    new_join.into(),
+                    Condition {
+                        conjunctions: other_condition,
+                    },
+                )
+            }
+            JoinType::LeftSemi | JoinType::LeftAnti | JoinType::Unspecified => unreachable!(),
+        }
+    }
+
```
Comment (likg227) [inline src/frontend/src/optimizer/rule/apply_join.rs:359]: ditto

**Review by likg227 — APPROVED** [review]
LGTM

File: src/frontend/src/optimizer/rule/apply_join.rs (hunk 1 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/rule/apply_join.rs:236]: Yes and done.

**Pull request merged**